text/0000-portability-lint.md

```diff
@@ -0,0 +1,574 @@
+  for a greater degree of portability, for example allowing their code to work
+  in the `no_std` ecosystem. Depending on the library, this may entail a
+  significant amount of work.
+
+There's a fundamental tradeoff here. On the one hand, we want Rust libraries to
+be as portable as possible. On the other hand, achieving *maximal* portability
+can be a big burden for library authors.  Our approach so far has been to
+identify "mainstream platform assumptions", as mentioned above, and *guide* code
+to work on all mainstream platforms by default; by convention, such portability
+is the default expectation of libraries on crates.io. This RFC formalizes that
+approach in a deeper way.
+
+An important point: while we can expect library authors who are striving for
+portability to test their code on a variety of target platforms, we can't make
+that assumption for the average library. In other words, **if we want to guide all
+Rust code toward at least mainstream portability, we will need to do so in a way
+that doesn't require actually compiling and testing for all mainstream
+scenarios**.
+
+# Detailed design
+[design]: #detailed-design
+
+## The basic idea
+
+The core problem we want to solve is:
+
+- We want to make non-mainstream APIs available in their natural location,
+  e.g. as inherent methods directly on standard library types.
+
+- We want to have some kind of "speed bump" before using such APIs, so that
+  users realize that they may be giving up mainstream portability.
+
+- We want to do this *without* requiring testing on platforms that lack the API.
+
+Let's take a concrete example: the `as_raw_fd` method. We'd like to provide this
+API as an inherent method on things like files. But it's not a "mainstream" API;
+it only works on Unix. If you tried to use it and compiled your code on Windows,
+you would discover the problem right away, since the API would not be available
+due to `cfg`. But if you were only testing on Linux, you might never notice,
+since the API is available there.
+
+**The basic idea of this RFC is to provide an additional layer of checking on
+top of the existing `cfg` system, to avoid usage of an API *accidentally working*
+because you happen to be compiling for a given target platform**. This checking
+is performed through a new **portability lint**, which warns when invoking APIs
+marked with `cfg` unless you've explicitly acknowledged the portability
+implications. We'll see how you do that in a moment.
+
+Going back to our example, we'd like to define methods on `File` like:
+
+```rust
+impl File {
+    #[cfg(unix)]
+    fn as_raw_fd(&self) -> RawFd { ... }
+
+    #[cfg(windows)]
+    fn as_raw_handle(&self) -> RawHandle { ... }
+}
+```
+
+If you attempted to call `as_raw_fd`, when compiling on Unix you'd get a warning
+from the portability lint that you're calling an API not available on all
+mainstream platforms. There are basically three ways to react (all of which will
+make the warning go away):
```

Comment:
Probably worth making it clear that there's a fourth option of explicitly suppressing the lint in cases where the other three options won't work for whatever reason.